File: 290-word-pattern/word-pattern.kt
class Solution {
    fun wordPattern(pattern: String, s: String): Boolean {
        val t = s.split(" ");
        if(pattern.length != t.size) return false

        var map : MutableMap<Char,String> = mutableMapOf()

        for(i in 0 until pattern.length){
            if(map[pattern[i]] == null){
                if(map.containsValue(t[i])) return false
                map[pattern[i]] = t[i]
            }
            else{
                if(map[pattern[i]] != t[i]) return false
            }
        }

        return true
    }
}

File: 290-word-pattern/word-pattern.py
class Solution(object):
    def wordPattern(self, pattern, s):
        """
        :type pattern: str
        :type s: str
        :rtype: bool
        """
        if len(pattern) != len(s.split()):
            return False
        m = dict()
        for p , st in zip(pattern,s.split()):
            if p in m:
                if m[p] != st:
                    return False
            elif st in m.values():
                return False
            m[p] = st
        return True